// zeebe-node/dist/lib/SimpleLogger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZBSimpleLogger = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
/**
 * Simple logger for ZBClient
 */
const logger = (loglevel) => (logMessage) => {
    let message;
    try {
        const parsedMessage = JSON.parse(logMessage);
        const gRPC = parsedMessage.id === 'gRPC Channel' ? ' [gRPC Channel]:' : '';
        const taskType = parsedMessage.taskType
            ? ` [${parsedMessage.taskType}]`
            : '';
        const msg = typeof parsedMessage.message === 'object'
            ? JSON.stringify(parsedMessage.message)
            : parsedMessage.message;
        message = `| zeebe | ${gRPC}${taskType} ${loglevel}: ${msg}`;
    }
    catch (e) {
        message = logMessage;
    }
    const time = dayjs_1.default().format('HH:mm:ss.SSS');
    const err = new Error('Debug stack trace');
    const stack = err.stack.split('\n');
    // tslint:disable: no-console
    const loggers = {
        DEBUG: console.info,
        ERROR: console.error,
        INFO: console.info,
    };
    const logMethod = loggers[loglevel];
    const info = loglevel === 'DEBUG'
        ? `${time} ${message}\n${stack}`
        : `${time} ${message}`;
    logMethod(info);
};
exports.ZBSimpleLogger = {
    debug: logger('DEBUG'),
    error: logger('ERROR'),
    info: logger('INFO'),
};
